packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -107,34 +100,67 @@ export function useComboBoxState<T extends object>(props: ComboBoxProps<T>): Com
 
   let initialSelectedKeyText = collection.getItem(props.selectedKey)?.textValue;
   let initialDefaultSelectedKeyText = collection.getItem(props.defaultSelectedKey)?.textValue;
-  let [inputValue, setInputValue] = useControlledState(toString(props.inputValue) || initialSelectedKeyText, toString(props.defaultInputValue) || initialDefaultSelectedKeyText || '', onInputChange);
+  let [inputValue, setInputValue] = useControlledState(toString(props.inputValue), initialSelectedKeyText || toString(props.defaultInputValue) || initialDefaultSelectedKeyText || '', onInputChange);
 
-  let selectedKey = computeKeyFromValue(inputValue, collection);
+  let selectedKey = props.selectedKey || computeKeyFromValue(inputValue, collection);
   let selectedKeys = useMemo(() => selectedKey != null ? [selectedKey] : [], [selectedKey]);
 
-  let setSelectedKey = (key) => {
-    if (key !== selectedKey) {
-      let item = collection.getItem(key);
-      let itemText = item ? item.textValue : '';
-      setInputValue(itemText);
-      triggerState.setOpen(false);
+  let triggerState = useMenuTriggerState(props);
+  // Fires when user hits Enter or clicks
+  let setSelectedKey = useCallback((key) => {
```

Comment:
Still uncertain about this setSelectedKey + the useEffect below for when prop.selectedKey is updated + its interactions with onInputChange.